map/chart_generator.cpp

```diff
@@ -0,0 +1,275 @@
+    // No need break or return here.
+  case MapStyleDark:
+    return agg::rgba8(255, 230, 140, 255);
+  case MapStyleLight:
+  case MapStyleClear:
+  case MapStyleMerged:
+    return agg::rgba8(30, 150, 240, 255);
+  }
+}
+
+agg::rgba8 GetCurveColor(MapStyle mapStyle)
+{
+  switch (mapStyle)
+  {
+  case MapStyleCount:
+    LOG(LERROR, ("Wrong map style param."));
+    // No need break or return here.
+  case MapStyleDark:
+    return agg::rgba8(255, 230, 140, 20);
+  case MapStyleLight:
+  case MapStyleClear:
+  case MapStyleMerged:
+    return agg::rgba8(30, 150, 240, 20);
+  }
+}
+}  // namespace
+
+namespace maps
+{
+bool NormalizeChartData(vector<double> const & distanceDataM,
+                        feature::TAltitudes const & altitudeDataM, size_t resultPointCount,
+                        vector<double> & uniformAltitudeDataM)
+{
+  double constexpr kEpsilon = 1e-6;
+
+  if (distanceDataM.size() != altitudeDataM.size())
+  {
+    LOG(LERROR, ("Altitude and distance data have different size."));
+    return false;
+  }
+
+  if (!is_sorted(distanceDataM.cbegin(), distanceDataM.cend()))
+  {
+    LOG(LERROR, ("Route segment distances are not sorted."));
+    return false;
+  }
+
+  if (distanceDataM.empty() || resultPointCount == 0)
+  {
+    uniformAltitudeDataM.clear();
+    return true;
+  }
+
+  auto const calculateAltitude = [&](double distFormStartM) {
+    if (distFormStartM <= distanceDataM.front())
+      return static_cast<double>(altitudeDataM.front());
+    if (distFormStartM >= distanceDataM.back())
+      return static_cast<double>(altitudeDataM.back());
+
+    auto const lowerIt = lower_bound(distanceDataM.cbegin(), distanceDataM.cend(), distFormStartM);
+    size_t const nextPointIdx = distance(distanceDataM.cbegin(), lowerIt);
+    ASSERT_LESS(0, nextPointIdx, ("distFormStartM is greater than 0 but nextPointIdx == 0."));
+    size_t const prevPointIdx = nextPointIdx - 1;
+
+    if (my::AlmostEqualAbs(distanceDataM[prevPointIdx], distanceDataM[nextPointIdx], kEpsilon))
+      return static_cast<double>(altitudeDataM[prevPointIdx]);
+
+    double const k = (altitudeDataM[nextPointIdx] - altitudeDataM[prevPointIdx]) /
+                     (distanceDataM[nextPointIdx] - distanceDataM[prevPointIdx]);
+    return static_cast<double>(altitudeDataM[prevPointIdx]) +
+           k * (distFormStartM - distanceDataM[prevPointIdx]);
+  };
+
+  double const routeLenM = distanceDataM.back();
+  uniformAltitudeDataM.resize(resultPointCount);
+  double const stepLenM = resultPointCount <= 1 ? 0.0 : routeLenM / (resultPointCount - 1);
+
+  for (size_t i = 0; i < resultPointCount; ++i)
+    uniformAltitudeDataM[i] = calculateAltitude(static_cast<double>(i) * stepLenM);
+
+  return true;
+}
+
+bool GenerateYAxisChartData(uint32_t height, double minMetersPerPxl,
+                            vector<double> const & altitudeDataM, vector<double> & yAxisDataPxl)
+{
+  yAxisDataPxl.clear();
+  if (altitudeDataM.empty())
+    return true;
```

Comment:
Please, move yAxisDataPxl.clear() to the block where you return true. Otherwise it's possible to return false but modify the result argument.